File: system/GenericController.go
package system

import (
	"action"
	"fmt"
	"os"
	"strings"
)

type GenericController struct {
	Id           string
	Namespace    string
	RelAddr      string
	AgeClustered bool
	Model        [4]*GenericModel
}

func (me *GenericController) NewGenericModel(op action.Action, fieldset []*Field) *GenericModel {
	var model = &GenericModel{
		Name:         me.Id,
		Suffix:       "",
		AgeClustered: me.AgeClustered,
		Fieldset:     fieldset,
	}

	if op == action.NONE {
		return model
	}

	if op == action.GET {
		model.Suffix = "List"
	}

	me.Model[op] = model

	return me.Model[op]
}

func (me *GenericController) getDatasetImpl() string {
	if me.AgeClustered {
		return "\n\t\t$request = new Request();\n\t\treturn _GET::find([\"conditions\" => \"_age = :_age:\", \"bind\" => [\"_age\" => $request->getHeader(\"Age\")]]);\n\t"
	} else {
		return "\n\t\treturn _GET::find();\n\t"
	}
}

func (me *GenericController) Publish(addr string) error {
	var byt []byte
	var e error
	if byt, e = os.ReadFile("system\\GenericController.php"); e != nil {
		return e
	}

	if me.Model[action.GET] == nil {
		me.Model[action.GET] = me.Model[action.SET]
	}

	if me.Model[action.SET] == nil {
		me.Model[action.SET] = me.Model[action.GET]
	}

	// var updbutton = "{ \\\"data\\\": null, \\\"className\\\": \\\"dt-center editor-edit\\\", \\\"defaultContent\\\": \\\"<i class=\\\\\\\"fa fa-pencil\\\\\\\" />\\\", \\\"orderable\\\": false }"
	// var rembutton = "{ \\\"data\\\": null, \\\"className\\\": \\\"dt-center editor-delete\\\", \\\"defaultContent\\\": \\\"<i class=\\\\\\\"fa fa-trash\\\\\\\" />\\\", \\\"orderable\\\": false }"
	var model = "use %s\\Models\\%s as Enum_%d;\n"
	var enum = "\n\t\t$enum_%d = Enum_%d::find([\"order\" => \"%s\"]);\n\t\t$this->view->setVar(\"enum_%d\", $enum_%d);\n\t"
	var getter = "\n\t\t$model->%s = $this->getViewStateValue(\"%s\");"

	var lib = ""
	var optionset = ""
	var modelreceiver = "\n\t\t$model = new _SET();\n\t\t$id = $this->getViewStateValue(\"_id\");\n\t\tif ($id != NULL) {\n\t\t\t$model = _SET::findFirst([\n\t\t\t\t\"conditions\" => \"_id = :_id:\",\n\t\t\t\t\"bind\" => [\"_id\" => $id]\n\t\t\t]);\n\t\t}\n"
	for i, field := range me.Model[action.SET].Fieldset {
		modelreceiver += fmt.Sprintf(getter, field.Name, field.Name)

		if field.Optionset != nil {
			lib += fmt.Sprintf(model, me.Namespace, field.Optionset.Name, i)
			optionset += fmt.Sprintf(enum, i, i, field.Optionset.Value, i, i)
		}
	}

	modelreceiver += "\n\t\treturn $model;"
	// rowformat = append(rowformat, updbutton)
	// rowformat = append(rowformat, rembutton)

	var php = string(byt)
	php = strings.Replace(string(php), "namespace System", fmt.Sprintf("namespace %s", me.Namespace), -1)
	php = strings.Replace(string(php), "use System\\Models\\GenericModel as _GET", fmt.Sprintf("use %s\\Models\\%s%s as _GET", me.Namespace, me.Model[action.GET].Name, me.Model[action.GET].Suffix), -1)
	php = strings.Replace(string(php), "use System\\Models\\GenericModel as _SET", fmt.Sprintf("use %s\\Models\\%s%s as _SET", me.Namespace, me.Model[action.SET].Name, me.Model[action.SET].Suffix), -1)
	php = strings.Replace(string(php), "use System\\Enums;", lib, -1)
	php = strings.Replace(string(php), "abstract class GenericController", fmt.Sprintf("class %sController", me.Id), -1)

	php = strings.Replace(string(php), "abstract function getViewState(): _SET;", fmt.Sprintf("private function getViewState(): _SET {%s\n\t}\n\t", modelreceiver), -1)
	php = strings.Replace(string(php), "abstract function getDataset(): Object;", fmt.Sprintf("private function getDataset(): Object {%s}\n\t", me.getDatasetImpl()), -1)
	php = strings.Replace(string(php), "abstract function dropdown(): void;", fmt.Sprintf("private function dropdown(): void {%s}", optionset), -1)

	var arr1 []string
	var arr2 []string
	if me.Model[action.SET].Suffix == "List" {
		arr1 = strings.Split(php, "\n\tpublic function newAction()")
		arr2 = strings.Split(arr1[1], "} // new")
		php = arr1[0] + arr2[1]

		arr1 = strings.Split(php, "\n\tpublic function getAction()")
		arr2 = strings.Split(arr1[1], "} // get")
		php = arr1[0] + arr2[1]

		arr1 = strings.Split(php, "\n\tpublic function putAction()")
		arr2 = strings.Split(arr1[1], "} // put")
		php = arr1[0] + arr2[1]

		arr1 = strings.Split(php, "\n\tpublic function remAction()")
		arr2 = strings.Split(arr1[1], "} // rem")
		php = arr1[0] + arr2[1]
	}

	byt = []byte(php)
	if e = os.WriteFile(fmt.Sprintf("%s\\controllers\\%sController.php", addr, me.Id), byt, 0644); e != nil {
		return e
	}

	var view *GenericView
	for i, model := range me.Model {
		var op = action.Action(i)
		if model == nil || (op == action.GET && model.Suffix == "") {
			continue
		}

		if e = model.Publish(me.Namespace, addr); e != nil {
			return e
		}

		if op == action.GET {
			continue
		}

		view = NewGenericView(me.Id, model.Fieldset)
		if e = view.Publish(addr, op); e != nil {
			return e
		}
	}

	return e
}

File: system/GenericController.php
<?php
namespace System\Controllers;

use System\Models\GenericModel as _GET;
use System\Models\GenericModel as _SET;
use System\Enums;
use Phactory\Auth\SecurityUsers as Authorization;
use Phalcon\Http\Response;
use Phalcon\Http\Request;
use Phalcon\Mvc\Controller;

abstract class GenericController extends Controller {

	private Request $request;
	private Response $response;

	public function initialize() {
		// Create a request and response instance
		$this->request = new Request();
		$this->response = new Response();
	}

	abstract function getViewState(): _SET;
	abstract function getDataset(): Object;
	abstract function dropdown(): void;

	private function getListId(): string {
		return sprintf("_%s_list", strtolower(_GET::getName()));
	}

	private function getViewStateValue(string $key): ?string {
		$value = $this->request->getPost($key);
		return !empty($value)? $value: NULL;
	}

	private function getToolbarRights(): string {
		$ctrl = strtolower(_GET::getName());
		$from = $this->request->getHeader("From");
		return sprintf(
			"{\"new\": %b, \"put\": %b, \"rem\": %b}",
			is_callable([$this, "GETACTION"]) && Authorization::allow($ctrl, "new", $from),
			is_callable([$this, "PUTACTION"]) && Authorization::allow($ctrl, "put", $from),
			is_callable([$this, "REMACTION"]) && Authorization::allow($ctrl, "rem", $from)
		);
	}

	public function openAction(): Response {
		if (!Authorization::allow(strtolower(_GET::getName()), "open", $this->request->getHeader("From"))) {
			$this->response->setStatusCode(401, "Unauthorized");
			return $this->response;
		}

		$this->view->disable();

		$content = sprintf(
			"{ \"id\": \"%s\", \"src\": \"%s\", \"fieldset\": %s, \"rowformat\": %s, \"headset\": [\"%s\"], \"rights\": %s }",
			$this->getListId(),
			strtolower(_GET::getName()),
			(is_callable([$this, "PUTACTION"])? _SET::getSerializedFieldset(): _GET::getSerializedFieldset()),
			_GET::getRowFormat(),
			implode("\", \"", _GET::getHeadset()),
			$this->getToolbarRights(),
		);

		$this->response->setContent($content);
		$this->response->setContentType("application/json");
		return $this->response;
	}

	public function listAction(): Response {
		if (!Authorization::allow(strtolower(_GET::getName()), "list", $this->request->getHeader("From"))) {
			$this->response->setStatusCode(401, "Unauthorized");
			return $this->response;
		}

		$this->view->disable();

		$dataset = $this->getDataset();
		$headset = _GET::getHeadset();
		$body = [];
		foreach ($dataset as $data) {
			$row = [];
			$j = 0;
			foreach ($headset as $i => $head) {
				$row[] = sprintf("\"%s\": \"%s\"", $j++, str_replace("\"", "\\\"", $data->$i));
			}

			$body[] = sprintf("{ %s }", implode(", ", $row));
		}

		$content = sprintf("{ \"data\": [%s] }", implode(", ", $body));
		$this->response->setContent($content);
		$this->response->setContentType("application/json");
		return $this->response;
	}

	public function newAction(): Response {
		$this->dropdown();
		$this->response->setContent($this->view->getRender(_SET::getName(), "set"));
		return $this->response;
	} // new

	public function getAction(): Response {
		if (!Authorization::allow(strtolower(_SET::getName()), "get", $this->request->getHeader("From"))) {
			$this->response->setStatusCode(401, "Unauthorized");
			return $this->response;
		}

		$this->dropdown();
		$this->view->setVar("identity", "none");

		$model = _SET::findFirst([
			"conditions" => "_id = :_id:",
			"bind" => ["_id" => $this->request->getPost("_id")]
		]);

		$headset = _SET::getHeadset();
		foreach ($headset as $i => $head) {
			$this->view->setVar($i, $model->$i);
		}

		$this->response->setContent($this->view->getRender(_SET::getName(), "set"));
		return $this->response;
	} // get

	public function putAction(): Response {
		if (!Authorization::allow(strtolower(_SET::getName()), "put", $this->request->getHeader("From"))) {
			$this->response->setStatusCode(401, "Unauthorized");
			return $this->response;
		}

		$this->view->disable();

		$model = $this->getViewState();
		$ok = false;
		try {
			$ok = $model->save();
		} catch (\Exception $e) {
			header("Warning: ".$e->getMessage());
		}

		if ($ok === false) {
			$this->response->setStatusCode(409, "Conflict");
		} else {
			$this->response->setStatusCode(200, "OK");
		}

		$this->response->setContent("{}");
		$this->response->setContentType("application/json");
		return $this->response;
	} // put

	public function remAction(): Response {
		if (!Authorization::allow(strtolower(_SET::getName()), "rem", $this->request->getHeader("From"))) {
			$this->response->setStatusCode(401, "Unauthorized");
			return $this->response;
		}

		$this->view->disable();

		$model = _SET::find(sprintf("_id in ('%s')", implode("', '", $this->request->getPost("_id"))));

		$ok = false;
		try {
			$ok = $model->delete();
		} catch (\Exception $e) {
			header("Warning: ".$e->getMessage());
		}

		if ($ok === false) {
			$this->response->setStatusCode(405, "Method Not Allowed");
		} else {
			$this->response->setStatusCode(200, "OK");
		}

		$this->response->setContent("{}");
		$this->response->setContentType("application/json");
		return $this->response;
	} // rem

}
?>